File: frontend/src/types/payments.ts
import type { RegosCurrencyOption } from "@/types/settings";

export type PaymentDirection = "income" | "outcome";

export type PaymentDocument = {
  id: number;
  code: string;
  date: number;
  amount: number | null;
  category_id: number | null;
  category_name: string | null;
  payment_type_id: number | null;
  payment_type_name: string | null;
  partner_id: number | null;
  partner_name: string | null;
  firm_id: number | null;
  firm_name: string | null;
  attached_user_id: number | null;
  attached_user_name: string | null;
  exchange_rate: number | null;
  description: string | null;
  performed: boolean;
  deleted_mark: boolean;
  payment_direction: PaymentDirection | null;
  currency: RegosCurrencyOption | null;
};

export type PaymentsListResponse = {
  documents: PaymentDocument[];
  next_offset: number;
  total: number;
};

export type PaymentCreateRequest = {
  firm_id: number;
  partner_id: number;
  direction: PaymentDirection;
  payment_type_id: number;
  amount: number;
  exchange_rate?: number | null;
  category_id?: number | null;
  description?: string | null;
  date?: number | null;
};

export type PaymentEditRequest = {
  firm_id?: number | null;
  partner_id?: number | null;
  payment_type_id?: number | null;
  amount?: number | null;
  exchange_rate?: number | null;
  category_id?: number | null;
  description?: string | null;
  date?: number | null;
};

export type PaymentMutationResponse = {
  row_affected: number;
};

File: backend/app/schemas/payments.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.settings import RegosCurrencyOption


class PaymentDocument(BaseModel):
    id: int
    code: str
    date: int
    amount: float | None = None
    category_id: int | None = None
    category_name: str | None = None
    payment_type_id: int | None = None
    payment_type_name: str | None = None
    partner_id: int | None = None
    partner_name: str | None = None
    firm_id: int | None = None
    firm_name: str | None = None
    attached_user_id: int | None = None
    attached_user_name: str | None = None
    exchange_rate: float | None = None
    description: str | None = None
    performed: bool = False
    deleted_mark: bool = False
    payment_direction: Literal["income", "outcome"] | None = None
    currency: RegosCurrencyOption | None = None


class PaymentsListResponse(BaseModel):
    documents: list[PaymentDocument]
    next_offset: int = 0
    total: int = 0


class PaymentCreateRequest(BaseModel):
    firm_id: int = Field(ge=1)
    partner_id: int = Field(ge=1)
    direction: Literal["income", "outcome"]
    payment_type_id: int = Field(ge=1)
    amount: float = Field(gt=0)
    exchange_rate: float | None = Field(default=None, gt=0)
    category_id: int | None = Field(default=None, ge=1)
    description: str | None = Field(default=None, max_length=1000)
    date: int | None = Field(default=None, ge=0)


class PaymentEditRequest(BaseModel):
    firm_id: int | None = Field(default=None, ge=1)
    partner_id: int | None = Field(default=None, ge=1)
    payment_type_id: int | None = Field(default=None, ge=1)
    amount: float | None = Field(default=None, gt=0)
    exchange_rate: float | None = Field(default=None, gt=0)
    category_id: int | None = Field(default=None, ge=1)
    description: str | None = Field(default=None, max_length=1000)
    date: int | None = Field(default=None, ge=0)


class PaymentMutationResponse(BaseModel):
    row_affected: int = Field(ge=0)
